### ISSUE TITLE
Add keystroke to toggle Focus Highlight

### ISSUE THREAD

**Issue by Qchristensen** [nvaccess/nvda#11405]
### Is your feature request related to a problem? Please describe.

There are times when it would be useful to toggle focus highlight.  Being able to assign a keystroke for this would make it easier to do for users who want this (we've had a few requests since adding Focus Highlight to NVDA in 2019.3).

### Describe the solution you'd like

A keystroke (not assigned by default), so that users can set this in Input Gestures if preferred.

### Describe alternatives you've considered

It can currently be done from the options, although the fewest keystrokes to do it is:
- NVDA+control+k (to open keyboard settings)
- control+shift+tab (to go up one category to Vision)
- spacebar (to toggle "Enable highlighting"
- enter (to close settings).

**Comment by CyrilleB79** [comment]
I think a gesture for each type of cursor highlighting would also be required. This would mean 4 unassigned gestures corresponding to the 4 checkbox of the Focus Highlight section in the vision panel.

**Comment by feerrenrut** [comment]
>I think a gesture for each type of cursor highlighting would also be required. 

I don't think that is necessary. Further it will cause inconsistency as more options are added, i don't think it is a good idea to have scripts for every configuration option, only things that need to be toggled regularly in certain situations. I expect most users of focus highlighting don't fiddle with the preferences, rather just want to turn it on and off.

**Comment by CyrilleB79** [comment]
> > I think a gesture for each type of cursor highlighting would also be required.
> 
> (...) i don't think it is a good idea to have scripts for every configuration option, only things that need to be toggled regularly in certain situations.

I agree on this general point.

> I expect most users of focus highlighting don't fiddle with the preferences, rather just want to turn it on and off.

Let me share my personal experience. I do not usually use focus highlight. I use it only rarely for test to be able to see where the navigator object is located as well as the system focus. Given my low vision, I prefer to have just one highlighter at a time. But I want to have:
* often: no highlighter
* rarely only the object navigator highlighter
* rarely only the system focus highlighter
Last but not least, I think that the options in the Vision category specifically should have a shortcut since opening the Vision setting window triggers a focus change and thus makes the magnified view move (e.g. using Windows Magnifier).

**Comment by XLTechie** [comment]
Could it be a rotary control, that cycles from off, through the various 
highlight options, back to off?

**Comment by CyrilleB79** [comment]
No a rotary control is clearly not adapted in this case since each option is not exclusive. Thus the rotary should allow to walk through each option but also through each combination of option... The various options are checkboxes, not radio buttons.

**Comment by CyrilleB79** [comment]
Thinking again at it, we may implement a unique keystroke for the whole highlight feature in a first time. If then people still need to have a keystroke to modify each cursor option, they will make it known and this could be discussed in a second time.

Implementing a unique keystroke for the highlighting feature would mean the following:
* the "Enable Highlighting" checkbox should become a binary checkbox so that the feature could simply be toggled. Presently, this checkbox is a 3-state checkbox.
* The 3 other checkboxes should be disabled (greyed out) if the main checkbox is unchecked and enabled (set focusable) if the main checkbox is checked. Checking or unchecking the main checkbox should not control anymore if the 3 other checkboxes are checked or unchecked.
* The 3 checkboxes specifying the type of highlight should not be all unchecked. We could issue an error message:
  * when unchecking one of these checkboxes if the 2 other ones are already checked
  or
  * when the user press OK or Apply to validate the setting panel.
  I do not know which of these 2 solutions would be the most suitable.

Cc @leonardder since I think you have implemented the highlighter feature.

**Comment by feerrenrut** [comment]
@CyrilleB79 while @leonardder provided the initial implementation, I developed this particular UI.

The three state check box is intended to convey that the highlighter is enabled, but not all things are highlighted. This was a bit of a UI trade off, but it was done under the assumption that the most common use case would be to turn on all options at once. The trade-off being that if a user does have a preference other than "all", it is lost when the highlighter is disabled.

Still, as a quick way to meet the needs of most, I'd suggest starting with a command that simply acts as if the enabled checkbox is toggled (enabling all / disabling all), then follow up by investigating how to preserve the preferences without affecting this use case.

**Comment by eilanabenish** [comment]
It would be very helpful to have shortcut keys for the different options of focus highlight .
For enable or disable focus visible 
for enable or disable system focus 
for enable or disable the navigator object 
for enable or disable browse mode cursor 
I have found that the CTRL + NVDA and the numbers are available for now at least in my NVDA and the add-ons I am running .
Thanks in advance

**Comment by CyrilleB79** [comment]
@eilanabenish wrote:
> It would be very helpful to have shortcut keys for the different options of focus highlight .

I agree with you. Anyway, let's do it step by step: as requested by @feerrenrut let's first include a command to toggle globally Visual Highlight. Then when it is merged in NVDA, let's discuss about the possibility to toggle each option individually.

> I have found that the CTRL + NVDA and the numbers are available for now at least in my NVDA and the add-ons I am running 

In any case, either for global Visual Highlight toggle or for individual option toggle command, I would recommand not to assign it by default. Indeed, Visual Highlight is important only for a minority of NVDA users since I guess most of the users are too much blind to need this feature. Users interested by this feature may always assign a shortcut key to the input gesture manager.

If useful, for my personal use, I have a prototype of this feature: [visualHighlighterToggler.py](https://raw.githubusercontent.com/CyrilleB79/NVDAScripts/master/globalPlugins/visualHighlighterToggler.py). To be downloaded (context menu / save as) and pasted in the globalPlugins of your scratchpad directory, provided you are familiar with this.

It is not fully working when NVDA starts up, but after 2 key press, seems OK. I need to investigate it. Then I will propose it to NVAccess, provided that they accept such new features.

**Comment by feerrenrut** [comment]
I'm happy to accept a PR for this work, and happy for the UI layout to be changed but please try to preserve the following use-cases:
- Use GUI to enable all highlight contexts
- Use GUI to disable all highlight contexts
- Use GUI to easily enable a single highlighter context

In addition, new use-cases I think are important:
- Use GUI to disable then re-enable without loosing the users preferred highlight contexts 

Maybe this means there should be a single on/off switch and a buttons:
- Check all
- Check none

The tricky part comes with "check None", if no highlight contexts are active, then the highlighter is effectively off.

**Comment by Qchristensen** [comment]
The current UI could be a little confusing for those who ONLY want one or two options enabled, AND want to be able to turn the feature on and off at will preserving that state.  I must admit I have always either had it on or off..

While I think there could likely be improvement in the overall design, I think the most beneficial first step would be beneficial to have a gesture to toggle the main visual highlight > enable highlighting feature on and off.

We do still get requests for this  (I'm going to update the title of this issue to reflect the feature name)
